### PR TITLE
add rari yag

### PR DESCRIPTION
**Pull request by dmelotik**
Rari Yield Aggregator Subgraph

https://rari.capital/
https://docs.rari.capital/yag/

TODOs:
- [x] Get manifest working
- [x] Start with deposits/withdrawals
- [x] finish the rest of the entities for RYPT
- [x] get snapshots working
- [x] get inputTokenBalances working
- [x] calculate revenue/fees
- [x] fix revenue negative issue
- [x] make helper methods pool indistinguishable
- [x] add usdc pool
- [x] add dai pool
- [x] add ether pool
- [x] get eth price
- [x] update `YieldAggregator` to update properly for all pools
- [x] aggregate price for `outputTokenPrice`
- [x] add in `PricePerShare`
- [x] revenues seem too large
- [x] remove unnecessary logs
- [x] final testing

See README.md `Known Issues` for issues and weird things

### CODE REVIEW

File: subgraphs/rari-yag-v1/src/common/utils/constants.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+export const ETHER_VAULT_NAME = "Rari Ether Pool";
+export const ETHER_VAULT_SYMBOL = "REPT"; // REPT = rariether pool token
+
+/////////////////////////////
+///// Protocol Specific /////
+/////////////////////////////
+
+export const RARI_DEPLOYER = "0xb8f02248d53f7edfa38e79263e743e9390f81942"; // using as "protocol address" b/c no factory contract
+export const PROTOCOL_NAME = "Earn";
```
Comment (this-username-is-taken) [inline subgraphs/rari-yag-v1/src/common/utils/constants.ts:224]: Is this what they call this product? If so, we can change it to "Rari Earn" and "rari-earn"

Comment (this-username-is-taken) [reply subgraphs/rari-yag-v1/src/common/utils/constants.ts:224]: On their website, it seems to be called Vaults. Maybe "Rari Vaults"?

Comment (dmelotik) [reply subgraphs/rari-yag-v1/src/common/utils/constants.ts:224]: thoughts on:
`name` = "Rari Vaults v1"
`slug` = "rari-vaults-v1"

Comment (dmelotik) [reply subgraphs/rari-yag-v1/src/common/utils/constants.ts:224]: on their ui looks like they label them "Vaults", but they are technically v1 vaults with v2 coming soon

Comment (this-username-is-taken) [reply subgraphs/rari-yag-v1/src/common/utils/constants.ts]: yeah, that sounds good!

Comment (dmelotik) [reply subgraphs/rari-yag-v1/src/common/utils/constants.ts]: fixed

File: subgraphs/rari-yag-v1/src/mappings/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    let contract = RariYieldFundManager.bind(Address.fromString(YIELD_VAULT_MANAGER_ADDRESS));
+    let withdrawalFee = getOrCreateVaultFee(VaultFeeType.WITHDRAWAL_FEE, vaultAddress);
+    tryWithdrawalFee = contract.try_getWithdrawalFeeRate();
+    tryPerfFee = contract.try_getInterestFeeRate();
+
+    if (!tryWithdrawalFee.reverted) {
+      // only update fee if it was not reverted
+      withdrawalFee.feePercentage = tryWithdrawalFee.value
+        .toBigDecimal()
+        .div(exponentToBigDecimal(DEFAULT_DECIMALS - 2));
```
Comment (harsh9200) [inline subgraphs/rari-yag-v1/src/mappings/helpers.ts:179]: Hello @dmelotik , can you avoid hard coding `DEFAULT_DECIMALS`, you should use the `inputToken.decimals` instead.

File: subgraphs/rari-yag-v1/src/common/prices/prices.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+// calculate prices
```
Comment (harsh9200) [inline subgraphs/rari-yag-v1/src/common/prices/prices.ts:1]: can you remove this file?

Comment (dmelotik) [reply subgraphs/rari-yag-v1/src/common/prices/prices.ts]: done

File: subgraphs/rari-yag-v1/src/mappings/helpers.ts (hunk 2 above)
Comment (dmelotik) [reply subgraphs/rari-yag-v1/src/mappings/helpers.ts:179]: Good catch I did change some of these to `inputToken.decimals`

However there are some places where the BigInt is offset by 1e18 on the contract-level. And here is the contract call for this one: https://github.com/Rari-Capital/rari-yield-pool-contracts/blob/master/contracts/RariFundManager.sol#L954

File: subgraphs/rari-yag-v1/README.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+`Vault TVL`
+
+### Total Revenue USD
+
+Sum across all Vaults:
+
+`Total Interest Accrued`
+
+Note: This currently excludes Liquidations
```
Comment (this-username-is-taken) [inline subgraphs/rari-yag-v1/README.md:17]: No need for this line since it doesn't have liquidations

File: subgraphs/rari-yag-v1/README.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+Count of Unique Addresses which have interacted with the protocol via any transaction
+
+`Deposits`
+
+`Withdraw`
+
+### Reward Token Emissions Amount
+
+To be added
```
Comment (this-username-is-taken) [inline subgraphs/rari-yag-v1/README.md:53]: Change to `Not Applicable`

File: subgraphs/rari-yag-v1/README.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+### Protocol Controlled Value
+
+To be added
```
Comment (this-username-is-taken) [inline subgraphs/rari-yag-v1/README.md:57]: Change to `Not Applicable`

File: subgraphs/rari-yag-v1/src/common/utils/constants.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+export const MS_PER_YEAR = DAYS_PER_YEAR.times(new BigDecimal(BigInt.fromI32(24 * 60 * 60 * 1000)));
+export const BLOCKS_PER_YEAR = BigDecimal.fromString("2102400");
+
+///////////////////////////
+//// Rari Input Tokens ////
+///////////////////////////
+
+// Note: Deposit/Withdrawal events return the currency as an indexed string
+// so we need to create a mapping between the 'currencyCode' to Token
+
```
Comment (this-username-is-taken) [inline subgraphs/rari-yag-v1/src/common/utils/constants.ts:144]: do you know how the indexed string is created/hashed?

**Comment by this-username-is-taken** [comment]
Great work! Just a couple minor fixed and we are good to go!

**Comment by this-username-is-taken** [comment]
Can you work with @steegecs to add the CI/CD flow to this subgraph as well (and turn it on by default)?

**Comment by steegecs** [comment]
> Can you work with @steegecs to add the CI/CD flow to this subgraph as well (and turn it on by default)?

Just let me know if you have any questions or want to meet @harsh9200

File: subgraphs/rari-yag-v1/src/common/utils/constants.ts (hunk 7 above)
Comment (dmelotik) [reply subgraphs/rari-yag-v1/src/common/utils/constants.ts:144]: added commenting, see https://medium.com/mycrypto/understanding-event-logs-on-the-ethereum-blockchain-f4ae7ba50378 for more info

**Comment by dmelotik** [comment]
directory structure updated for deployment scripts @this-username-is-taken @steegecs

**Pull request merged**